src/core/net/ip6.cpp

```diff
@@ -1179,19 +1177,27 @@ Error Ip6::HandleDatagram(Message &aMessage, MessageOrigin aOrigin, const void *
         {
             receive = true;
         }
-        else if (!header.GetDestination().IsLinkLocal())
+        else if ((aOrigin != kFromThreadNetif) || !header.GetDestination().IsLinkLocal())
         {
-            forwardThread = true;
-        }
-        else if (aOrigin != kFromThreadNetif)
-        {
-            forwardThread = true;
-        }
+            if (header.GetDestination().IsLinkLocal())
+            {
+                forwardThread = true;
+            }
+            else if (IsOnLink(header.GetDestination()))
+            {
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_DUA_NDPROXYING_ENABLE
+                forwardThread = ((aOrigin == kFromHostDisallowLoopBack) ||
+                                 !Get<BackboneRouter::Manager>().ShouldForwardDuaToBackbone(header.GetDestination()));
+#else
+                forwardThread = true;
+#endif
+            }
+            else if (Get<ThreadNetif>().RouteLookup(header.GetSource(), header.GetDestination()) == kErrorNone)
+            {
+                forwardThread = true;
+            }
 
-        if (forwardThread && !ShouldForwardToThread(messageInfo, aOrigin))
-        {
-            forwardThread = false;
-            forwardHost   = true;
+            forwardHost = !forwardThread;
```

Comment:
The `diff` here seems more complex than it actually is, so let me describe it as two back-to-back changes:

As original code we had: 
```c++
        if (Get<ThreadNetif>().HasUnicastAddress(header.GetDestination()))
        {
            receive = true;
        }
        else if (!header.GetDestination().IsLinkLocal())
        {
            forwardThread = true;
        }
        else if (aOrigin != kFromThreadNetif)
        {
            forwardThread = true;
        }

        if (forwardThread && !ShouldForwardToThread(messageInfo, aOrigin))
        {
            forwardThread = false;
            forwardHost   = true;
        }
```

We can convert this to:
```c++
            receive = true;
        }
        else if ((aOrigin != kFromThreadNetif) || !header.GetDestination().IsLinkLocal())
        {
            forwardThread = ShouldForwardToThread(messageInfo, aOrigin);
            forwardHost   = !forwardThread;
        }
```
Basically combining the two `if()` checks that were setting `forwardThread = true` and replace the `if (forwardThread && !ShouldForwardToThread(messageInfo, aOrigin))` check by setting the `forwardThread` value directly and `forwardHost` as opposite of the `forwardThread`.

Next change is to inline the `ShouldForwardToThread()` implementation:
```c++
bool Ip6::ShouldForwardToThread(const MessageInfo &aMessageInfo, MessageOrigin aOrigin) const
{
    bool shouldForward = false;

    if (aMessageInfo.GetSockAddr().IsMulticast() || aMessageInfo.GetSockAddr().IsLinkLocal())
    {
        shouldForward = true;
    }
    else if (IsOnLink(aMessageInfo.GetSockAddr()))
    {
#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_DUA_NDPROXYING_ENABLE
        shouldForward = ((aOrigin == kFromHostDisallowLoopBack) ||
                         !Get<BackboneRouter::Manager>().ShouldForwardDuaToBackbone(aMessageInfo.GetSockAddr()));
#else
        OT_UNUSED_VARIABLE(aOrigin);
        shouldForward = true;
#endif
    }
    else if (Get<ThreadNetif>().RouteLookup(aMessageInfo.GetPeerAddr(), aMessageInfo.GetSockAddr()) == kErrorNone)
    {
        shouldForward = true;
    }

    return shouldForward;
}
```

We can note that 
- `aMessageInfo.GetSockAddr()` is the same as `header.GetDestination()`  and
- `aMessageInfo.GetPeerAddr()` is the same as `header.GetSource()`.

Also we don't need the `aMessageInfo.GetSockAddr().IsMulticast()` check anymore since we are already under `else` block where we checked that destination is unicast.

Reply:
LGTM. Thanks for the additional explanation!